### PR TITLE
[13.0] [ADD] stock_picking_product_barcode_report: New module that allows you to select how many labels print for each stock_move_lines

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa TT25883

This module allows you to select how many labels print for each product of a picking. You can see an example of the wizard and the result in next gif:

As you see in last example the format of the barcode is GS1-128, you can select the default format on the configuration. If you set the camp as blank it will be working with the Odoo formats by default.

Please @sergio-teruel @carlosdauden review this.

### CODE REVIEW

**Comment by lmignon** [comment]
@CarlosRoca13 Thank you for this proposal. It would be nice to be able to select the report format from the print wizard with the default value set to the company's default. ...

**Comment by CarlosRoca13** [comment]
@lmignon What you suggest is to select the format of the barcode on the wizard? or select the paperformat?

**Comment by lmignon** [comment]
> @lmignon What you suggest is to select the format of the barcode on the wizard? or select the paperformat?

I suggest to select theformat of the barcode with default defined by https://github.com/OCA/stock-logistics-barcode/pull/300/files#diff-2a2a39ed07c5da1815a017a4cd2c41e9R21

File: stock_picking_product_barcode_report/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Carlos Roca <carlos.roca@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# -*- coding: utf-8 -*-
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: You can remove this line...

File: stock_picking_product_barcode_report/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Carlos Roca <carlos.roca@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: Add https...
Be careful with copy - paste

File: stock_picking_product_barcode_report/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+from odoo.addons.web.controllers.main import ReportController
+
+
+class ReportController(ReportController):
+    @http.route(
+        ["/report/barcode", "/report/barcode/<type>/<path:value>"],
+        type="http",
+        auth="public",
+    )
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: If you do not change any thing you only need  @http.route()

File: stock_picking_product_barcode_report/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    def report_barcode(
+        self,
+        type,  # pylint: disable=redefined-builtin
+        value,
+        width=600,
+        height=100,
+        humanreadable=0,
+        quiet=1,
+    ):
+        barcode_format = request.env.company.barcode_default_format
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: This value must be taken from type definition...

File: stock_picking_product_barcode_report/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        if barcode_format == "gs1_128":
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: if type == "gs1_128"

File: stock_picking_product_barcode_report/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            Gs1_128 = barcode.get_barcode_class("gs1_128")
+            gs1_182 = Gs1_128(str(value), writer=SVGWitoutTextWriter())
+            image = gs1_182.render()
+            # image.show()
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py]: remove this line

File: stock_picking_product_barcode_report/controllers/main.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            type,
+            value,
+            width=width,
+            height=height,
+            humanreadable=humanreadable,
+            quiet=quiet,
+        )
+
+
+class SVGWitoutTextWriter(SVGWriter):
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/controllers/main.py:40]: Please add docstring explaining why yo need this class

File: stock_picking_product_barcode_report/readme/CONFIGURE.rst [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#. Go to 'Settings > Inventory > Barcode format' and select 'Display GS1_128 format for
+   barcodes'.
+#. Go to 'Inventory > Configuration > Operation Types' and active the 'Detailed operations'
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/readme/CONFIGURE.rst]: This step is not mandatory...

File: stock_picking_product_barcode_report/readme/USAGE.rst [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#. Create a new transfer with a product with 'Tracking -> By Lots' and 'Barcode' with a
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/readme/USAGE.rst]: Does the module work without lots???

File: stock_picking_product_barcode_report/report/report_label_barcode.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <data>
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/report/report_label_barcode.xml]: You can remove data tag

File: stock_picking_product_barcode_report/report/report_label_barcode_template.xml [hunk 11]
```diff
@@ -0,0 +1,88 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <template id="report_label_barcode">
+        <t t-call="web.basic_layout">
+            <div class="page text-justify">
+                <t t-foreach="doc.label_qty" t-as="i">
+                    <div class="text-center" t-if="doc.product_id.barcode">
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/report/report_label_barcode_template.xml]: Why not only create lines if they have barcodes?? If it you do not need this condition

File: stock_picking_product_barcode_report/report/report_label_barcode_template.xml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                            t-if="doc.env.user.company_id.barcode_default_format == 'gs1_128'"
+                        >
+                            <t t-if="doc.move_line_id.lot_id">
+                                <img
+                                    t-att-src="'/report/barcode/?type=%s&amp;value=%s&amp;width=%s&amp;height=%s' % ('gs1_128', '02' + doc.product_id.barcode.zfill(14) + '10' + doc.move_line_id.lot_id.name, 600, 100)"
+                                    style="width:100%;height:70px;margin: 0px 0px -25px 0px"
+                                    alt="Barcode"
+                                />
+                                <span
+                                    t-esc="'(02)' + doc.product_id.barcode.zfill(14) + '(10)' + doc.move_line_id.lot_id.name"
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/report/report_label_barcode_template.xml]: create a variable to store this and you can save a lot code lines.
Also you can store type in a variable...

File: stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    move_line_id = fields.Many2one("stock.move.line", "Move")
+
+
+class WizStockBarcodeSelectionPrinting(models.TransientModel):
+    _name = "stock.picking.print"
+    _description = "Wizard to select how many barcodes have to be printed"
+
+    @api.model
+    def default_get(self, fields):
+        ctx = self.env.context
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py]: ctx = self.env.context.copy()

File: stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        if len(ctx.get("active_ids", list())) > 1:
+            raise UserError(_("You may print the labels for just a picking"))
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py]: Why??? I want to print por more than one picking...

File: stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        res = super().default_get(fields)
+        if ctx.get("active_id") and ctx.get("active_model") == "stock.picking":
+            picking = self.env["stock.picking"].browse(ctx.get("active_id"))
+            if picking.exists():
+                res.update({"picking_id": picking.id})
+        return res
+
+    picking_id = fields.Many2one("stock.picking")
```
Comment (sergio-teruel) [inline stock_picking_product_barcode_report/wizard/stock_barcode_selection_printing.py]: You can change by picking_ids and we can process more than one picking

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@sergio-teruel changes done

**Comment by sergio-teruel** [comment]
Please add a boolean field to product for avoid print labels...
And add the field into wizard for change format like as @lmignon suggests

**Comment by CarlosRoca13** [comment]
@sergio-teruel @lmignon Changes done!

**Comment by CarlosRoca13** [comment]
Here you have an example of the result
captura

**Comment by CarlosRoca13** [comment]
Here you have an example of the result

**Comment by pedrobaeza** [comment]
Shouldn't the quantity of labels be defaulted to the quantity done or initial demand in the picking?

**Comment by CarlosRoca13** [comment]
That's a possibility, another possibility is like it is right now because we have no info about how the picking will arrive

**Comment by pedrobaeza** [comment]
OK, I let you decide.

**Comment by CarlosRoca13** [comment]
I add one change for avoid a bug on non admin users

**Comment by manuelcalerosolis** [comment]
@CarlosRoca13 congratulations on this useful module.

I miss a column to represent the lot number in the selection of the number of labels to print.

image

Is a backport to version 12.0 planned?

Regards

**Review by manuelcalerosolis — APPROVED** [review]
Functional test ok!

**Comment by CarlosRoca13** [comment]
I add the Lot Number on the wizard. Actually the backport to 12.0 is not planned, but probably it will be later. @manuelcalerosolis

**Review by sergio-teruel — APPROVED** [review]

**Pull request merged**